### PR TITLE
[refactor] Move output files to a common results/ folder

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

This PR changes the artifact output directory to this structure:
```
results/ 
    {run_id}/
        configuration.yaml - the configuration (after CLI overrides) used for this run.
        {behavior_name}/
            Tensorboard files
            snapshot.ckpt (and .pb) - model checkpoints and intermediate output files
        {behavior_name}.csv - CSV log
        {behavior_name}.nn - Barracuda Output
    run_logs/
        timers.json - timers file
```

In addition, we now output the RunOptions object as a YAML file `configuration.yaml`. When combined with PR #3815, the resulting YAML can then be used with `mlagents-learn` to re-run the same run again.

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

https://docs.google.com/document/d/1uyk5JVNevfWy2DmqVHy19lcDzyQCNppPxkPOgJcQ7z4/edit#heading=h.5cr9b5kpfi98

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [x] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by ervteng** [comment]
@anupambhatnagar - Added you to the review for visibility as it will require a corresponding change to the CI pipeline.

**Comment by harperj** [comment]
@ervteng you're right that this will break the daily CI runs.  I think there are 3 places in the CI we'll need to update:
- results reporter
- script for updating NN files
- the runner script's barracuda inference logic

One of us could take a stab if that would be useful.

Unrelatedly, I think the structure you listed in the PR description is a little off -- all files should be nested under the `run-id` folder.  I'd also suggest we might just get rid of the `run_logs` folder and move those files into the general `run-id` folder.

**Comment by ervteng** [comment]
@harperj I've actually already made the CI changes in this PR: https://github.com/Unity-Technologies/ml-agents-cloud/pull/105 but definitely it needs a looksee from one of you to make sure I didn't break anything. Seems to work when I tested it. 

I think getting rid of the `run-log` (`summaries`) in the design doc could work, as there will only be two files in there (the timers.json and eventually the Player.Log). The only time it will make sense to have a separate `run-log` is when running many environments, as each will have their own Player.log (so Player-1.log, Player-2.log, etc.) and clutter up the {run-id} directory.

File: ml-agents/mlagents/trainers/trainer_controller.py [hunk 1]
```diff
@@ -119,16 +117,16 @@ def _export_graph(self):
...
-                os.makedirs(model_path)
+            if not os.path.exists(output_path):
+                os.makedirs(output_path)
         except Exception:
             raise UnityEnvironmentException(
                 "The folder {} containing the "
                 "generated model could not be "
                 "accessed. Please make sure the "
-                "permissions are set correctly.".format(model_path)
+                "permissions are set correctly.".format(output_path)
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer_controller.py:129]: minor: while you're changing this, can you update to an f-string?

File: ml-agents/mlagents/trainers/learn.py [hunk 2]
```diff
@@ -320,26 +321,29 @@ def run_training(run_seed: int, options: RunOptions) -> None:
...
     with hierarchical_timer("run_training.setup"):
-        model_path = f"./models/{options.run_id}"
+        base_path = "results"
+        write_path = os.path.join(base_path, options.run_id)
         maybe_init_path = (
-            f"./models/{options.initialize_from}" if options.initialize_from else None
+            os.path.join(base_path, options.run_id) if options.initialize_from else None
         )
-        summaries_dir = "./summaries"
+        run_logs_dir = os.path.join(write_path, "run_logs")
```
Comment (harperj) [inline ml-agents/mlagents/trainers/learn.py:329]: You could use `pathlib` for some of these operations.  For example..

```python
base_path = Path('./results')
write_path = base_path / options.run_id
maybe_init_path = base_path / options.run_id if options.initialize_from else None
run_logs_dir = base_path / run_logs
run_logs_dir.mkdir(exist_ok=True)
```

Not a big deal, but I do think it is a little nicer than the `os` and `glob` equivalents.

File: ml-agents/mlagents/trainers/learn.py [hunk 3]
```diff
@@ -398,11 +400,26 @@ def run_training(run_seed: int, options: RunOptions) -> None:
...
+    except FileNotFoundError:
+        logger.warning(
+            f"Unable to save configuration to {run_options_path}. Make sure the directory exists"
+        )
 
 
-def write_timing_tree(summaries_dir: str, run_id: str) -> None:
-    timing_path = f"{summaries_dir}/{run_id}_timers.json"
+def write_timing_tree(output_dir: str) -> None:
+    timing_path = f"{output_dir}/timers.json"
```
Comment (harperj) [inline ml-agents/mlagents/trainers/learn.py:422]: minor: join paths here, to deal with cross-platform compat and output_dir possibly having a trailing slash

File: ml-agents/mlagents/trainers/trainer/trainer.py [hunk 4]
```diff
@@ -42,9 +42,8 @@ def __init__(
         self.brain_name = brain_name
         self.run_id = run_id
         self.trainer_parameters = trainer_parameters
-        self.summary_path = trainer_parameters["summary_path"]
         self._threaded = trainer_parameters.get("threaded", True)
-        self._stats_reporter = StatsReporter(self.summary_path)
+        self._stats_reporter = StatsReporter(brain_name)
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer/trainer.py:46]: Does this / how does this change the behavior of the StatsWriters?

File: ml-agents/mlagents/trainers/trainer_util.py [hunk 5]
```diff
@@ -83,9 +79,8 @@ def initialize_trainer(
 
     :param trainer_config: Original trainer configuration loaded from YAML
     :param brain_name: Name of the brain to be associated with trainer
-    :param summaries_dir: Directory to store trainer summary statistics
     :param run_id: Run ID to associate with this training run
-    :param model_path: Path to save the model
+    :param output_path: Path to save the model and save summary statistics
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer_util.py:83]:
```suggestion
    :param output_path: Path to save the model and summary statistics
```

File: ml-agents/mlagents/trainers/trainer_util.py [hunk 6]
```diff
@@ -102,9 +97,8 @@ def initialize_trainer(
         )
 
     trainer_parameters = trainer_config.get("default", {}).copy()
-    trainer_parameters["summary_path"] = str(run_id) + "_" + brain_name
-    trainer_parameters["model_path"] = "{basedir}/{name}".format(
-        basedir=model_path, name=brain_name
+    trainer_parameters["output_path"] = "{basedir}/{name}".format(
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer_util.py:100]: minor: join paths using `os.path.join` or similar

File: docs/Training-ML-Agents.md [hunk 7]
```diff
@@ -64,16 +64,17 @@ for a sample execution of the `mlagents-learn` command.
 Regardless of which training methods, configurations or hyperparameters you
 provide, the training process will always generate three artifacts:
 
-1. Summaries (under the `summaries/` folder): these are training metrics that
+1. Summaries (under the `results/<run-identifier>/<behavior-name>` folder):
+   these are training metrics that
```
Comment (andrewcoh) [inline docs/Training-ML-Agents.md:68]:
```suggestion
   These are training metrics that
```

File: ml-agents/mlagents/trainers/trainer/trainer.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer/trainer.py:46]: Previously the StatsReporter saved everything to the directory specified here, and `summary_path = {run_id}_{brain_name}` (no actual path, `summaries/` was assigned on construction of the StatsWriter in `learn.py`.). So changing to just `brain_name` puts it in `results/{brain_name}`.

File: ml-agents/mlagents/trainers/learn.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/learn.py:329]: Tried this, liked it, then realized Tensorflow's utilities don't support Path objects, and backed out as we'd have a bunch of Path to `str` conversions that are easy to mess up. Another thing to do once we convert to PyTorch...

Comment (harperj) [reply ml-agents/mlagents/trainers/learn.py:329]: OK -- no worries.  Let's just refactor to use it when it makes sense

**Pull request merged**